### ISSUE TITLE
Mismatched parentheses in command substitutions?

### ISSUE THREAD

**Issue by jakwings** [fish-shell/fish-shell#2876]
fish from git commit 35cee1e

```fish
echo (echo hi #hi)
```

```
> echo (echo hi #hi)
hi
```

The script above runs successfully, and the one below fails:

```fish
echo (echo hi #hi()
```

```
> echo (echo hi #hi()
Unexpected end of string, parenthesis do not match
fish:  echo (echo hi #hi()
```

**Comment by jakwings** [comment]
Two more examples:

```fish
echo (echo hi #()
         #)
```

```
> echo (echo hi #()
         #)
hi
```

```fish
echo (echo hi #(
         )
```

```
> echo (echo hi #(
         )
Unexpected end of string, parenthesis do not match
fish:  echo (echo hi #(
            ^
```

**Comment by faho** [comment]
"Honey, the parser's borked again!"

Yeah, I can reproduce. Nice catch!

**Comment by krader1961** [comment]
So what is the expected behavior and where is it documented? I can't find anything in the user facing documentation that even mentions `#` as introducing a comment. My expectation for the first example is the right-paren is part of the comment and thus ignored. Which means the second example should fail in exactly the same way since the second left-paren would also be ignored since it's part of the comment.

**Comment by jakwings** [comment]
I agree comment should be allowed in command substitutions, though we rarely see a multi-line command substitution.

**Comment by krader1961** [comment]
> I agree comment should be allowed in command substitutions...

@jakwings: Are you saying that a command substitution context should limit the scope of a comment? That is, should

```
echo (echo hi #there()
```
treat the second left-paren as part of the comment since it is grammatically limited to the outer command substitution? That is, as equivalent to

```
echo (echo hi)
```

Or are you agreeing with me that the comment introduction char, `#`, should behave like the equivalent syntax in languages such as Python and C++ where it causes every subsequent char up to the end of the line to be ignored?

**Comment by jakwings** [comment]
@krader1961 The latter. Because it is more consistent with comments outside the substitution.

At first I was not sure if comments are allowed there, or if the right-paren cannot stay in another line.
